meta: split CHANGELOG into two files

TheAlphaNerd opened the pull request:
The unfortunate has happened, our CHANGELOG is now over 1 MB and cannot
be viewed on github. This commit breaks the CHANGELOG into two files
so that we can continue to show our changes rendered in the github UI.

Closes: https://github.com/nodejs/node/issues/5533

It is very likely that we don't want to use `CHANGELOG-ARCHIVE.md` as a filename... this is just to get the ball rolling

claudiorodriguez commented:
LGTM, and +1 as dealing with huge changelogs can be painful when you just want to look at the recent changes. However, might be getting a bit ahead of myself here, but are we just going to keep on piling stuff onto the "archive" file, or are we going to go with something like `CHANGELOG-YYYY-MM-DD.md`, splitting when a determined filesize is reached?

silverwind commented:
May I suggest moving the archived file into `doc` to prevent further pollution of the root directory?

TheAlphaNerd commented:
@silverwind totally up for that... I want to wait to hear if anyone wants to follow a specific naming convention and I'll fix that up

ChALkeR commented:
Previous considerations: https://github.com/nodejs/node/issues/2810, https://github.com/nodejs/node/issues/5533.

jasnell commented:
LGTM with moving archive doc into docs dir

evanlucas commented:
Won't this break links in a lot of blog posts?

TheAlphaNerd commented:
We can alternatively take out everything from before v0.10 and that will give us a stopgap to decide what we want to do.

@evanlucas  at the very least it won't break any modern websites

bnoordhuis commented:
> It is very likely that we don't want to use CHANGELOG-ARCHIVE.md as a filename...

A project I used to work on had the last 1-1.5 years in CHANGELOG with everything before that split per year (CHANGELOG.2008, CHANGELOG.2009, etc.)

I agree that they should be kept out of the top-level directory.  Perhaps add a note at the bottom of the CHANGELOG that older entries can be found in doc/.

TheAlphaNerd commented:
I've updated the commit.

Everything older than a year is now in `doc/CHANGELOG.ARCHIVE.md`

I have added a link at the end of the changelog to the archive

jasnell commented:
@TheAlphaNerd ... Just a thought: how difficult do you think it would be to at least keep the headers of the removed releases and include a link to the new location? If folks have links to the changelog for specific releases then doing so may alleviate some of the pain of broken links.

evanlucas commented:
+1 to what @jasnell sqid

TheAlphaNerd commented:
updated. The caveat at the bottom now lists all the change logs that can be found in the archive.

It is worth mentioning that the CHANGELOG is now at 700k. We'll hit the limit again soon enough and I don't think including all the headers will scale. That being said I think it is a good step for now

jasnell commented:
I'm wondering if we shouldn't move to having the CHANGELOG.md only show the current changelogs for the active releases and move all previous changelogs to individual archive files. It would make things a bit more difficult but if each release has it's own changelog file then the links should always remain stable.

TheAlphaNerd commented:
@jasnell I like that idea quite a bit. How would that affect porting the changelog between release stream?

Would each stream have only the latest release in it's changelog and master would have the latest for all actively maintained channels?

Fishrock123 commented:
Imo, we should do archives in `doc/`.

I disagree with having just a separate "archive" file though. These ought to be separated by year, but I have not have the time to do it quite yet. (Some older entries are ordered by-version and not by-date.)

--

> Won't this break links in a lot of blog posts?

yes. We should consider how we go forward with this.

Perhaps we should have the root changelog only have heading links to the year-sorted versions?

jbergstroem commented:
How about splitting up changes in majors? (0.12.x, 4.x, 5.x, etc)

Fishrock123 commented:
@jbergstroem We discussed that previously and iirc people didn't like the idea. It's harder to tell when what happened.

claudiorodriguez commented:
After reading up on it, +1 on having archives in `doc/`separated by year.

TheAlphaNerd commented:
since we can't seem to come to an exact agreement, and the changelog is currently broken for viewing on github. Are people ok landing this the way it is an iterating what we do with the archive?

I'd like people to be able to read the changelog when v6 drops

bnoordhuis commented:
Sounds good.  LGTM.

silverwind commented:
LGTM, let's get this in now!

mscdex commented:
FWIW this should have had the `doc:` prefix instead of `meta:`

Fishrock123 commented:
> Imo, we should do archives in `doc/`.

> I disagree with having just a separate "archive" file though. These ought to be separated by year, but I have not have the time to do it quite yet. (Some older entries are ordered by-version and not by-date.)

--

> > Won't this break links in a lot of blog posts?

> yes. We should consider how we go forward with this.

> Perhaps we should have the root changelog only have heading links to the year-sorted versions?

Do I need to add a new `blocked` or `-1` label??

Fishrock123 commented:
Ok at least the links "work", but they redirect to "content" that is in no way useful...

Fishrock123 commented:
Fine, I'll spend some of my weekend fixing this too.

TheAlphaNerd commented:
@Fishrock123 this was discussed in https://github.com/nodejs/node/pull/6337#issuecomment-214802583

The changelog was broken in production on our release day and we were having multiple issues opened to let us know. While this change is not where we would like it to be, it was sufficient in causing the minimal breakage while fixing the immediate problem that we were facing. 

I'm more than happy to do the grunt work of getting this where it needs to be if we can reach a consensus of what needs to be done.

jasnell commented:
+1 @TheAlphaNerd. There are a couple of workable approaches that would
work. By year is one way, but I think I'd much prefer a separate file per
major. That's a bit more work but manageable. Happy to help!
On Apr 29, 2016 5:50 PM, "Myles Borins" <notifications@github.com> wrote:

> @Fishrock123 <https://github.com/Fishrock123> this was discussed in #6337
> (comment)
> <https://github.com/nodejs/node/pull/6337#issuecomment-214802583>
>
> The changelog was broken in production on our release day and we were
> having multiple issues opened to let us know. While this change is not
> where we would like it to be, it was sufficient in causing the minimal
> breakage while fixing the immediate problem that we were facing.
>
> I'm more than happy to do the grunt work of getting this where it needs to
> be if we can reach a consensus of what needs to be done.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/nodejs/node/pull/6337#issuecomment-215917195>
>
